File: android/app/src/main/java/com/aureliussecureface/frameprocessors/FaceDetectorPlugin.kt
package com.aureliussecureface.frameprocessors

import android.media.Image
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.face.FaceLandmark
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy
import kotlin.math.PI

class FaceDetectorPlugin(
    proxy: VisionCameraProxy,
    options: Map<String, Any>?
) : FrameProcessorPlugin() {

    private val detectorOptions = FaceDetectorOptions.Builder()
        .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
        .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
        .setContourMode(FaceDetectorOptions.CONTOUR_MODE_ALL)
        .build()

    private val detector = FaceDetection.getClient(detectorOptions)

    override fun callback(frame: Frame, params: Map<String, Any>?): Any? {
        val mediaImage: Image = frame.image

        // --- Conversion sécurisée de l'orientation ---
        val orientationDegrees = when (val o = frame.orientation) {
            is Number -> (o.toDouble() * 180.0 / PI).toInt()
            is String -> o.toDoubleOrNull()?.times(180.0 / PI)?.toInt() ?: 0
            else -> 0
        }

        val image = InputImage.fromMediaImage(mediaImage, orientationDegrees)

        // Traitement avec ML Kit
        val task = detector.process(image)
        val faces = Tasks.await(task)

        val result = mutableListOf<Map<String, Any>>()
        for (face in faces) {
            val map = mutableMapOf<String, Any>()

            // Bounding box
            val bounds = mapOf(
                "top" to face.boundingBox.top.toDouble(),
                "left" to face.boundingBox.left.toDouble(),
                "width" to face.boundingBox.width().toDouble(),
                "height" to face.boundingBox.height().toDouble()
            )
            map["bounds"] = bounds

            // Angles
            map["rollAngle"] = face.headEulerAngleZ.toDouble()
            map["pitchAngle"] = face.headEulerAngleX.toDouble()
            map["yawAngle"] = face.headEulerAngleY.toDouble()

            // Landmarks
            val landmarks = mutableMapOf<String, Map<String, Double>>()
            
            face.getLandmark(FaceLandmark.LEFT_EYE)?.let {
                landmarks["leftEye"] = mapOf("x" to it.position.x.toDouble(), "y" to it.position.y.toDouble())
            }
            face.getLandmark(FaceLandmark.RIGHT_EYE)?.let {
                landmarks["rightEye"] = mapOf("x" to it.position.x.toDouble(), "y" to it.position.y.toDouble())
            }
            face.getLandmark(FaceLandmark.NOSE_BASE)?.let {
                landmarks["noseBase"] = mapOf("x" to it.position.x.toDouble(), "y" to it.position.y.toDouble())
            }
            face.getLandmark(FaceLandmark.MOUTH_LEFT)?.let {
                landmarks["mouthLeft"] = mapOf("x" to it.position.x.toDouble(), "y" to it.position.y.toDouble())
            }
            face.getLandmark(FaceLandmark.MOUTH_RIGHT)?.let {
                landmarks["mouthRight"] = mapOf("x" to it.position.x.toDouble(), "y" to it.position.y.toDouble())
            }
            face.getLandmark(FaceLandmark.MOUTH_BOTTOM)?.let {
                landmarks["mouthBottom"] = mapOf("x" to it.position.x.toDouble(), "y" to it.position.y.toDouble())
            }

            map["landmarks"] = landmarks

            // Contours
            val contoursMap = mutableMapOf<String, List<Map<String, Double>>>()
            val contourTypes = intArrayOf(
                com.google.mlkit.vision.face.FaceContour.FACE,
                com.google.mlkit.vision.face.FaceContour.LEFT_EYE,
                com.google.mlkit.vision.face.FaceContour.RIGHT_EYE,
                com.google.mlkit.vision.face.FaceContour.LEFT_EYEBROW_TOP,
                com.google.mlkit.vision.face.FaceContour.LEFT_EYEBROW_BOTTOM,
                com.google.mlkit.vision.face.FaceContour.RIGHT_EYEBROW_TOP,
                com.google.mlkit.vision.face.FaceContour.RIGHT_EYEBROW_BOTTOM,
                com.google.mlkit.vision.face.FaceContour.NOSE_BRIDGE,
                com.google.mlkit.vision.face.FaceContour.NOSE_BOTTOM,
                com.google.mlkit.vision.face.FaceContour.LEFT_CHEEK,
                com.google.mlkit.vision.face.FaceContour.RIGHT_CHEEK,
                com.google.mlkit.vision.face.FaceContour.UPPER_LIP_TOP,
                com.google.mlkit.vision.face.FaceContour.UPPER_LIP_BOTTOM,
                com.google.mlkit.vision.face.FaceContour.LOWER_LIP_TOP,
                com.google.mlkit.vision.face.FaceContour.LOWER_LIP_BOTTOM
            )

            val contourNames = arrayOf(
                "FACE", "LEFT_EYE", "RIGHT_EYE", "LEFT_EYEBROW_TOP", "LEFT_EYEBROW_BOTTOM",
                "RIGHT_EYEBROW_TOP", "RIGHT_EYEBROW_BOTTOM", "NOSE_BRIDGE", "NOSE_BOTTOM",
                "LEFT_CHEEK", "RIGHT_CHEEK", "UPPER_LIP_TOP", "UPPER_LIP_BOTTOM",
                "LOWER_LIP_TOP", "LOWER_LIP_BOTTOM"
            )

            for (i in contourTypes.indices) {
                face.getContour(contourTypes[i])?.let { contour ->
                    contoursMap[contourNames[i]] = contour.points.map { point ->
                        mapOf("x" to point.x.toDouble(), "y" to point.y.toDouble())
                    }
                }
            }
            map["contours"] = contoursMap

            // --- Passive Texture Analysis (Emergency Fix) ---
            // Calculate pixel variance in the face region as a fallback for TFLite
            val variance = calculateLaplacianVariance(mediaImage, face.boundingBox)
            map["textureAnalysis"] = mapOf(
                "variance" to variance,
                "moireDetected" to false // To be replaced with TFLite Moire detector
            )

            result.add(map)
        }

        return result
    }

    private fun calculateLaplacianVariance(image: Image, bounds: android.graphics.Rect): Double {
        val plane = image.planes[0] // Use Y plane for variance calculation
        val buffer = plane.buffer
        val width = image.width
        val height = image.height
        val rowStride = plane.rowStride
        val pixelStride = plane.pixelStride

        // Clamp bounds to image dimensions
        val left = bounds.left.coerceIn(0, width - 1)
        val top = bounds.top.coerceIn(0, height - 1)
        val right = bounds.right.coerceIn(0, width - 1)
        val bottom = bounds.bottom.coerceIn(0, height - 1)

        if (right <= left || bottom <= top) return 0.0

        var sum = 0.0
        var sumSq = 0.0
        var count = 0

        // Simple Laplacian operator: [0, 1, 0; 1, -4, 1; 0, 1, 0]
        for (y in top + 1 until bottom - 1) {
            for (x in left + 1 until right - 1) {
                val center = buffer.get(y * rowStride + x * pixelStride).toInt() and 0xFF
                val leftPixel = buffer.get(y * rowStride + (x - 1) * pixelStride).toInt() and 0xFF
                val rightPixel = buffer.get(y * rowStride + (x + 1) * pixelStride).toInt() and 0xFF
                val topPixel = buffer.get((y - 1) * rowStride + x * pixelStride).toInt() and 0xFF
                val bottomPixel = buffer.get((y + 1) * rowStride + x * pixelStride).toInt() and 0xFF

                val laplacian = (leftPixel + rightPixel + topPixel + bottomPixel - 4 * center).toDouble()
                sum += laplacian
                sumSq += laplacian * laplacian
                count++
            }
        }

        if (count == 0) return 0.0
        val mean = sum / count
        return (sumSq / count) - (mean * mean)
    }
}

File: ios/AureliusSecureFace/FaceDetectorPlugin.swift
import VisionCamera
import MLKitFaceDetection
import MLKitVision

@objc(FaceDetectorPlugin)
public class FaceDetectorPlugin: FrameProcessorPlugin {
  private static let options: FaceDetectorOptions = {
    let options = FaceDetectorOptions()
    options.performanceMode = .accurate
    options.landmarkMode = .all
    options.contourMode = .all
    return options
  }()

  private let faceDetector = FaceDetector.faceDetector(options: FaceDetectorPlugin.options)

  @objc
  public init(proxy: VisionCameraProxy, options: [AnyHashable: Any]?) {
    super.init(proxy: proxy, options: options)
  }

  public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
    let image = VisionImage(buffer: frame.buffer)
    image.orientation = frame.orientation

    var faces: [Face]
    do {
      faces = try faceDetector.results(in: image)
    } catch {
      return nil
    }

    var result: [[String: Any]] = []
    for face in faces {
      var map: [String: Any] = [:]

      // Bounding box
      map["bounds"] = [
        "top": face.frame.origin.y,
        "left": face.frame.origin.x,
        "width": face.frame.size.width,
        "height": face.frame.size.height
      ]

      // Angles
      map["rollAngle"] = face.headEulerAngleZ
      map["pitchAngle"] = face.headEulerAngleX
      map["yawAngle"] = face.headEulerAngleY

      // Landmarks
      var landmarks: [String: [String: CGFloat]] = [:]
      let landmarkTypes: [(String, FaceLandmarkType)] = [
        ("leftEye", .leftEye),
        ("rightEye", .rightEye),
        ("noseBase", .noseBase),
        ("mouthLeft", .leftMouth),
        ("mouthRight", .rightMouth),
        ("mouthBottom", .mouthBottom)
      ]

      for (name, type) in landmarkTypes {
        if let landmark = face.landmark(ofType: type) {
          landmarks[name] = ["x": landmark.position.x, "y": landmark.position.y]
        }
      }
      map["landmarks"] = landmarks

      // Contours
      var contoursMap: [String: [[String: CGFloat]]] = [:]
      let contourTypes: [(String, FaceContourType)] = [
        ("FACE", .face),
        ("LEFT_EYE", .leftEye),
        ("RIGHT_EYE", .rightEye),
        ("LEFT_EYEBROW_TOP", .leftEyebrowTop),
        ("LEFT_EYEBROW_BOTTOM", .leftEyebrowBottom),
        ("RIGHT_EYEBROW_TOP", .rightEyebrowTop),
        ("RIGHT_EYEBROW_BOTTOM", .rightEyebrowBottom),
        ("NOSE_BRIDGE", .noseBridge),
        ("NOSE_BOTTOM", .noseBottom),
        ("LEFT_CHEEK", .leftCheek),
        ("RIGHT_CHEEK", .rightCheek),
        ("UPPER_LIP_TOP", .upperLipTop),
        ("UPPER_LIP_BOTTOM", .upperLipBottom),
        ("LOWER_LIP_TOP", .lowerLipTop),
        ("LOWER_LIP_BOTTOM", .lowerLipBottom)
      ]

      for (name, type) in contourTypes {
        if let contour = face.contour(ofType: type) {
          contoursMap[name] = contour.points.map { ["x": $0.x, "y": $0.y] }
        }
      }
      map["contours"] = contoursMap

      // --- Passive Texture Analysis ---
      let variance = calculateLaplacianVariance(frame.buffer, bounds: face.frame)
      map["textureAnalysis"] = [
        "variance": variance,
        "moireDetected": false // Placeholder
      ]

      result.append(map)
    }

    return result
  }

  private func calculateLaplacianVariance(_ buffer: CMSampleBuffer, bounds: CGRect) -> Double {
    guard let imageBuffer = CMSampleBufferGetImageBuffer(buffer) else { return 0.0 }
    CVPixelBufferLockBaseAddress(imageBuffer, .readOnly)
    defer { CVPixelBufferUnlockBaseAddress(imageBuffer, .readOnly) }

    guard let baseAddress = CVPixelBufferGetBaseAddressOfPlane(imageBuffer, 0) else { return 0.0 }
    let width = CVPixelBufferGetWidth(imageBuffer)
    let height = CVPixelBufferGetHeight(imageBuffer)
    let bytesPerRow = CVPixelBufferGetBytesPerRowOfPlane(imageBuffer, 0)

    let left = Int(max(0, min(CGFloat(width - 1), bounds.origin.x)))
    let top = Int(max(0, min(CGFloat(height - 1), bounds.origin.y)))
    let right = Int(max(0, min(CGFloat(width - 1), bounds.origin.x + bounds.size.width)))
    let bottom = Int(max(0, min(CGFloat(height - 1), bounds.origin.y + bounds.size.height)))

    if right <= left || bottom <= top { return 0.0 }

    var sum: Double = 0.0
    var sumSq: Double = 0.0
    var count = 0

    let data = baseAddress.assumingMemoryBound(to: UInt8.self)

    for y in (top + 1)..<(bottom - 1) {
      for x in (left + 1)..<(right - 1) {
        let center = Double(data[y * bytesPerRow + x])
        let leftPixel = Double(data[y * bytesPerRow + (x - 1)])
        let rightPixel = Double(data[y * bytesPerRow + (x + 1)])
        let topPixel = Double(data[(y - 1) * bytesPerRow + x])
        let bottomPixel = Double(data[(y + 1) * bytesPerRow + x])

        let laplacian = leftPixel + rightPixel + topPixel + bottomPixel - 4 * center
        sum += laplacian
        sumSq += laplacian * laplacian
        count += 1
      }
    }

    if count == 0 { return 0.0 }
    let mean = sum / Double(count)
    return (sumSq / Double(count)) - (mean * mean)
  }
}
